### ISSUE TITLE
Reinstatement of of ssh tunnel

### ISSUE THREAD

**Issue by chrisburrell** [fishtown-analytics/dbt#1345]
Hi

I'm looking at : https://github.com/fishtown-analytics/dbt/issues/93 which suggests that once upon a time ssh-host would have been a valid property under a target of profiles.yml where a user can set up an ssh tunnel to connect to the data source.

I would love to be able to use this feature again. 

### Some context
In our case, the scenario is that our data is brought in by stitch in many schemas. I'd like to give local access to a data engineer to some of that data, but that redshift is secured using a data bastion sitting in front.

### Some issues we're trying to resolve
In our case, we can develop a lot of the scripts locally against postgres, however there are limitations:
* Speed is not greatly representative locally (anonymised production data / cut down data sets, etc.)
* Redhisft & Postgres differ in function names - we use jinja to abstract that away in the model files, but we can't test portions of jinja until we're in Redshift itself and using production data.

### What we would like to do
* Allow a local user to connect to Redshift through our data bastion. They have access to all the source models) but can only right to a 'user' schema (say 'chris').
* They can then test the actual DBT scripts per-se.

**Comment by drewbanin** [comment]
Thanks @chrisburrell - I'll have a look at why we removed this in the first place. Hopefully our next iteration of it can avoid whatever pitfalls we encountered last time around :)

**Comment by chrisburrell** [comment]
@drewbanin - Thanks for the easy workaround 
``ssh -l $BASTION_USER $BASTION_HOST -p 22 -N -C -L "5439:${REDSHIFT_HOST}:5439";``

Then use bash.rc or any other means to run it quickly/constantly

**Comment by drewbanin** [comment]
I don't think this is something we're going to want to add to dbt. It's easy enough to do on a user's machine, and the complexity (running a separate background thread/process) makes it not worth it. There are lots of good open source ssh tunnel tools out there that do a good job of this - closing this.

**Comment by dashirov** [comment]
This feature request is important enough to be implemented in dbt cloud, but apparently not important enough to help developers test in local profiles. rly.

**Comment by z3z1ma** [comment]
> This feature request is important enough to be implemented in dbt cloud, but apparently not important enough to help developers test in local profiles. rly.

Of course it needs to be done in dbt cloud. Locally just use the ssh command or better yet autossh where you can set it and forget it. Your reply isn't constructive and comes off as entitled. And developer worth his grain of salt knows how to open an ssh tunnel without dbt needing to take it on. 

Regards,

**Comment by jtcohen6** [comment]
Hey folks, I know we all have varying levels of experience with SSH-based connections, bastion hosts, etc. This confused me the first time I had to do it a few years ago.

We decided to keep this capability separate from `dbt-core` because it's not really specific to dbt. There are good tools for solving this problem, specific to the machine you're running on.

The basic idea, as I remember it:
- You register a public SSH key with the remote location, tied to a private key that lives on your machine
- You use a CLI tool (e.g. `ssh`, `autossh`) to "forward" a local port to the remote location (bastion host)
- In `profiles.yml`, instead of putting the host/port of a remote database, you put `localhost` and the number of the "forwarding" port
- Voila! Your connection is forwarded to the bastion host, authenticated via SSH, and passed along to the database

Does anyone have any good walkthroughs / blog posts with examples that they recommend when onboarding a new colleague to this sort of setup?

**Comment by litwak** [comment]
I just hacked my way through figuring this out and the steps listed by @jtcohen6 and @chrisburrell were very helpful for someone with zero experience in this realm who still needs to use dbt with a bastion host.  Here is specifically how I did this and some helpful resources I came across.  Hopefully others will find these examples helpful.

> You register a public SSH key with the remote location, tied to a private key that lives on your machine

Github has a helpful guide for how to do this:  https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent
Add keys to ~/.ssh/config: https://stackoverflow.com/questions/24392657/adding-an-rsa-key-without-overwriting
I also had to add `IgnoreUnknown UseKeychain` to ~/.ssh/config 

> You use a CLI tool (e.g. ssh, autossh) to "forward" a local port to the remote location (bastion host)

To forward the local port to the bastion host, save your user/bastion host/db host into environment variables.  I used Postgres so it looked like this.
`ssh -l $BASTION_USER $BASTION_HOST -p 22 -N -C -L "5432:${POSTGRES_HOST}:5432";`

> In profiles.yml, instead of putting the host/port of a remote database, you put localhost and the number of the "forwarding" port

Then in my ~/.dbt/profiles.yml looks includes this:
```
dev:
    type: postgres
    threads: 1
    host: localhost 
    port: 5432
    user: POSTGRES_USER
    pass: POSTGRES_PWD
    dbname: POSTGRES_DB_NAME
    schema: dbt_tmp
```

> Voila! Your connection is forwarded to the bastion host, authenticated via SSH, and passed along to the database
